src/TTImp/ProcessBuiltin.idr

```diff
@@ -0,0 +1,163 @@
+||| Checking a %builtin pragma is correct.
+-- If we get more builtins it might be wise to move each builtin to it's own file.
+module TTImp.ProcessBuiltin
+
+import Libraries.Data.Bool.Extra
+import Libraries.Data.NameMap
+import Data.List
+
+import Core.Core
+import Core.Context
+import Core.Context.Log
+import Core.Env
+import Core.Metadata
+import Core.TT
+import Core.UnifyState
+
+import TTImp.TTImp
+
+||| Get the return type.
+getRetTy : {vars : _} -> Term vars -> (vars ** Term vars)
+getRetTy (Bind _ _ _ tm) = getRetTy tm
+getRetTy tm = (vars ** tm)
+
+||| Get the first non-erased argument type.
+getFirstNETy : {vars : _} -> Term vars -> (vars ** Term vars)
+getFirstNETy (Bind _ _ b tm) = if isErased $ multiplicity b
+    then getFirstNETy tm
+    else (_ ** binderType b)
+getFirstNETy tm = (vars ** tm)
```

Comment:
Shouldn't this catchall be `Nothing` instead?

Reply:
It doesn't return a `Maybe`, are you suggesting it should return `Maybe`?
Also when it comes to the type system I basically just learnt it for this PR so you know better than me.

Reply:
Yes I think it should as if you cannot find a `NETy` it's better to explicitly
return `Nothing` than some potential garbage.

At the call site the check on non-erased args hopefully told you that there ought 
to be a value so if you get `Nothing` back you can raise an internal error. This
way if we have a broken invariant somewhere we'll get an error rather than silently
continuing with the wrong domain.